arroyo-api/src/pipelines.rs

```diff
@@ -339,35 +348,82 @@ pub async fn validate_pipeline(
         preview: false,
     };
 
-    let (mut program, _) = compile_sql(&sql, &auth_data, &client).await?;
-    optimizations::optimize(&mut program.graph);
+    // check the udfs
+    let mut controller = ControllerGrpcClient::connect(state.controller_addr.clone())
+        .await
+        .map_err(log_and_map)?;
 
-    let nodes = program
-        .graph
-        .node_weights()
-        .map(|node| PipelineNode {
-            node_id: node.operator_id.to_string(),
-            operator: format!("{:?}", node),
-            parallelism: node.clone().parallelism as u32,
-        })
+    let udf_strings = validate_pipeline_post
+        .udfs
+        .clone()
+        .unwrap_or(vec![])
+        .iter()
+        .filter(|&u| u.language == arroyo_rpc::types::UdfLanguage::Rust)
+        .map(|u| u.definition.clone())
         .collect();
-    let edges = program
-        .graph
-        .edge_references()
-        .map(|edge| {
-            let src = program.graph.node_weight(edge.source()).unwrap();
-            let target = program.graph.node_weight(edge.target()).unwrap();
-            PipelineEdge {
-                src_id: src.operator_id.to_string(),
-                dest_id: target.operator_id.to_string(),
-                key_type: edge.weight().key.to_string(),
-                value_type: edge.weight().value.to_string(),
-                edge_type: format!("{:?}", edge.weight().typ),
+
+    let check_udfs_resp = controller
+        .check_udfs(CheckUdfsReq { udfs: udf_strings })
+        .await
+        .map_err(log_and_map)?
+        .into_inner();
+
+    let udf_validation_result = if check_udfs_resp.result == ValidationResult::Error as i32 {
+        UdfsValidationResult {
+            udfs: None,
+            errors: Some(check_udfs_resp.errors),
+        }
+    } else {
+        UdfsValidationResult {
+            udfs: Some(validate_pipeline_post.udfs.unwrap_or(vec![])),
+            errors: None,
+        }
+    };
+
+    let pipeline_graph_validation_result = match compile_sql(&sql, &auth_data, &client).await {
```

Comment:
Does it make sense to validate the sql first, so that we can report errors back more quickly?

Reply:
Yup, these are now separate endpoints.